trio/_threads.py

```diff
@@ -70,6 +71,71 @@ class ThreadPlaceholder:
     name: str = attr.ib()
 
 
+# Types for the to_thread_run_sync message loop
+@attr.s(frozen=True, eq=False)
+class ThreadDone(Generic[RetT]):
+    result: outcome.Outcome[RetT] = attr.ib()
+
+
+@attr.s(frozen=True, eq=False)
+class Run(Generic[RetT]):
+    afn: Callable[..., Awaitable[RetT]] = attr.ib()
+    args: tuple[object, ...] = attr.ib()
+    context: contextvars.Context = attr.ib()
+    queue: stdlib_queue.SimpleQueue[outcome.Outcome[RetT]] = attr.ib(
+        init=False, factory=stdlib_queue.SimpleQueue
+    )
+
+    @disable_ki_protection
+    async def unprotected_afn(self) -> RetT:
+        coro = coroutine_or_error(self.afn, *self.args)
+        return await coro
+
+    async def run(self) -> None:
+        task = trio.lowlevel.current_task()
+        old_context = task.context
+        task.context = self.context.copy()
+        try:
+            await trio.lowlevel.cancel_shielded_checkpoint()
+            result = await outcome.acapture(self.unprotected_afn)
+            self.queue.put_nowait(result)
+        finally:
+            task.context = old_context
+            await trio.lowlevel.cancel_shielded_checkpoint()
+
+    async def run_system(self) -> None:
+        result = await outcome.acapture(self.unprotected_afn)
+        self.queue.put_nowait(result)
+
+
+@attr.s(frozen=True, eq=False)
+class RunSync(Generic[RetT]):
+    fn: Callable[..., RetT] = attr.ib()
+    args: tuple[object, ...] = attr.ib()
+    context: contextvars.Context = attr.ib()
+    queue: stdlib_queue.SimpleQueue[outcome.Outcome[RetT]] = attr.ib(
+        init=False, factory=stdlib_queue.SimpleQueue
+    )
+
+    @disable_ki_protection
+    def unprotected_fn(self) -> RetT:
+        ret = self.fn(*self.args)
+
+        if inspect.iscoroutine(ret):
+            # Manually close coroutine to avoid RuntimeWarnings
+            ret.close()
+            raise TypeError(
+                "Trio expected a sync function, but {!r} appears to be "
```

Comment:
nit: I think our usual documentation style is to spell out "synchronous function" but abbreviate "async function" for parity with the syntax

Reply:
This is really old text, and protected by a test, but on the other hand I'm happy to update them both!